Store ImageUploader attachments in Asset Manager as well as NFS

chrislo opened the pull request:
This PR switches the `ImageUploader` to use the `asset_manager_and_quarantined_file_storage` storage engine - ensuring that any new files that are uploaded using this uploader will be stored in Asset Manager. This is a precursor to migrating all of the existing files that use this uploader to Asset Manager and eventually removing them from NFS altogether. 

The change itself (in the first commit) is simple. Subsequent commits add some additional integration testing to make sure that Create, Update and Delete operations on a model that mounts this uploader (`Person`) work as expected. 

Note that the `ImageUploader` is configured to not remove files when they are replaced with new ones.

Fixes: https://github.com/alphagov/asset-manager/issues/325

chrislo commented:
In the case where a file is uploaded using ImageUploader with the same filename (but different contents) the `gds-api-adapter` will throw an `GdsApi::HTTPUnprocessableEntity` exception because the `Legacy url path is already taken`. 

I'm going to close this PR for now and re-open when I've fixed this issue.

floehopper commented:
Minor: There's a typo in the first commit note: "Assert Manager" -> "Asset Manager".

floehopper commented:
Minor: There's a typo in the commit note for "Integration test for replacing an ImageUploader image": "but I have varied that setting" -> "but I have verified that setting".

floehopper commented:
I think it would be useful to summarise the impact of the `remove_previously_stored_files_after_update` uploader config setting in the PR description. In particular I'd like to know whether there will be any difference from a URL point of view for updated assets stored on Whitehall NFS vs those stored in Asset Manager.

floehopper commented:
Review summary:

* I'm slightly unsure about the integration tests which only check the number of calls to the Asset Manager API, but maybe they're OK.
* I still find calls to `Mocha::Expectation#with` with a block argument hard to read. We've talked about addressing separately it in previous PRs, but this PR adds more occurrences.
* I think it would be useful to summarise the impact of the `remove_previously_stored_files_after_update` uploader config setting in the PR description. In particular I'd like to know whether there will be any difference for updated assets from a URL point of view for assets stored on Whitehall NFS vs those stored in Asset Manager.

I can't actually add a review until this is re-opened.

floehopper commented:
Review summary:

* I'm slightly unsure about the integration tests which only check the number of calls to the Asset Manager API, but maybe they're OK.
* I still find calls to Mocha::Expectation#with with a block argument hard to read. We've talked about addressing separately it in previous PRs, but this PR adds more occurrences.
* I think it would be useful to summarise the impact of the remove_previously_stored_files_after_update uploader config setting in the PR description. In particular I'd like to know whether there will be any difference for updated assets from a URL point of view for assets stored on Whitehall NFS vs those stored in Asset Manager.

I can't actually add a review until this is re-opened.

chrislo commented:
I've fixed the typos, refactored the `Mocha::Expectation#with` calls and added some more detail to the PR description. I can't reopen this PR as I've added new commits, so have opened #3602 instead.
